apps/aecore/test/aec_chain_sim.erl

```diff
@@ -373,39 +426,76 @@ terminate(_Reason, _Chain) ->
 code_change(_FromVsn, C, _Extra) ->
     {ok, C}.
 
-%% Called from the chain process
+
+%%%===================================================================
+%%% gen_server requests implementations
+%%%===================================================================
+
+new_account_(ForkId, Balance, #{forks := Forks} = Chain) ->
+    #{pubkey := PK} = KP = new_keypair(),
+    Acct = aec_accounts:new(PK, Balance),
+    #{ ForkId := #{blocks := [TopBlock|RestBlocks] = Blocks} = F } = Forks,
+    Trees = case trees(Blocks) of
+                {ok, Ts} ->
+                    Ts;
+                error ->
+                    aec_trees:new_without_backend()
+            end,
+    NewChain = insert_key_pair(KP, Chain),
+    NewTrees = aec_trees:set_accounts(Trees, aec_accounts_trees:enter(Acct, aec_trees:accounts(Trees))),
+    NewBlocks = [TopBlock#{trees => NewTrees}|RestBlocks],
+    {{ok, KP}, NewChain#{forks => Forks#{ForkId => F#{blocks => NewBlocks}}}}.
+
+get_balance_(ForkId, PK, Chain) ->
+    Trees = case trees(blocks(ForkId, Chain)) of
+                {ok, Ts} ->
+                    Ts;
+                error ->
+                    aec_trees:new_without_backend()
+            end,
+    case aec_accounts_trees:lookup(PK, aec_trees:accounts(Trees)) of
+        none -> 0;
+        {value, Acct} ->
+            aec_accounts:balance(Acct)
+    end.
+
 add_microblock_(ForkId, #{mempool := Pool} = Chain, Opts) ->
-    Txs = lists:reverse(Pool),
+    Txs = lists:reverse(Pool),  % because LIFO
     add_microblock_(ForkId, Txs, Chain#{mempool => []}, Opts).
 
 add_microblock_(ForkId, Txs, #{forks := Forks} = Chain, Opts) ->
     ?LOG("add_microblock(Txs = ~p", [Txs]),
     #{blocks := Blocks} = F = maps:get(ForkId, Forks),
-    #{hash := PrevHash, header := TopHdr} = hd(Blocks),
+    #{block := B} = hd(Blocks),
+    TopHdr = aec_blocks:to_header(B),
+    {ok, PrevHash} = aec_headers:hash_header(TopHdr),
     ?LOG("PrevHash = ~p", [PrevHash]),
     PrevKeyHash = aec_headers:prev_key_hash(TopHdr),
```

Comment:
@uwiger @radrow  It seems that the original chain simulator also sets the PrevKeyHash improperly.
This will always point to the genesis hash.
Let me demonstrate:
```
1. Genesis
2. Genesis <- Gen1Micro1(Prev: Genesis, PrevKey: Genesis) <- ...
3. Genesis <- ... <- Key1(Prev: Gen1MicroN, PrevKey: Genesis)
4. Genesis <- ... <- Key1(Prev: Gen1MicroN, PrevKey: Genesis) <- Gen2Micro1(Prev: Key1, PrevKey: **Genesis**
```

The correct code:
```
PrevKeyHash = case aec_headers:is_key(TopHdr) of true -> TopHdrHash; false -> aec_headers:prev_key_hash(TopHdr) end,
```